### DISCUSSION TITLE
Helper to write target factories

### DISCUSSION THREAD

**Discussion by liutiming** [ropensci/targets#288] [Help]
Hi @wlandau, I am wondering if you have a helper to write target factories from existing targets?

I feel it should be doable to have a script that parse the existing targets as a text file and
1. remove the commas
2. change target to target raw
3. add quotation marks, substitute and quote

**Comment by wlandau** [comment]
I am considering a new `tar_deparse()` function that would source `_targets.R` (just like `tar_manifest()`) and generate a bunch of `tar_target_raw()` calls using the internal settings of the actual target objects. But I am not committed yet because I do not fully understand the value it would add. I am not sure this adds much convenience for writing target factories, and there may be smoother workarounds when metaprogramming is needed.

**Reply by liutiming** [reply]
I think that will be very helpful. I am a bit confused which metaprogramming syntax should be used. The `tar_target_raw` function documentation says I should add `expression` to the existing target, but the [tar_stan_mcmc_rep.R](https://github.com/wlandau/stantargets/blob/4316142f6eb5359c5e6b952e657dd397c979765b/R/tar_stan_mcmc_rep.R) is using a combination of `quote` and `substitute`

**Reply by wlandau** [reply]
I do not think we can automate all the metaprogramming for something as involved as `tar_stan_mcmc_rep()`, but hopefully the updates to https://wlandau.github.io/targetopia/contributing.html#metaprogramming help people know when to use `quote()`, `deparse()`, and `substitute()`.

**Reply by liutiming** [reply]
Thanks @wlandau !

so if a function is like 
`tar_target_raw('tar_name',expression(f(a)), pattern = expression(map(a)), iteration = "list", format = "file")`

what do you use in place of `expression`? I am trying to re-learn metaprogramming as well but I am wondering what is a quick way to implement that in targets

**Reply by liutiming** [reply]
I think I am making some progress, but I am not sure why you used `quote`, not `substitute` on the top of the page 🤔 
`tar_target_raw("data", quote(read_data(file)), format = "fst_tbl", deployment = "main")`

**Reply by wlandau** [reply]
I want that first factory to be as simple as possible, so I am avoiding simple substitution. I want to generate the literal expression `read_data(file)`, but `substitute(read_data(file))` would output `read_data("value_of_file_argument")` because the megaprogramming is happening inside a function environment.

**Reply by liutiming** [reply]
Hmm I wouldn't say I understand perfectly. It will take some time. 

Is there a direct conversion from
`tar_target('tar_name', f(a), pattern = map(a), iteration = "list", format = "file")`
to 
`tar_target_raw('tar_name',expression(f(a)), pattern = expression(map(a)), iteration = "list", format = "file")` ?

where `a` is the name of another target name?

Sorry I do not want to sound lazy by seeking your advice. I am trying to learn this part well but I think it will be a quite steep learning curve

**Reply by wlandau** [reply]
`substitute()` can do this:

``` r
substitute(f(a), env = list(a = as.symbol("another_target_name")))
#> f(another_target_name)
```

<sup>Created on 2021-02-02 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Reply by liutiming** [reply]
ah sorry. What I meant was that `a` is already a target name. So you should not need to substitute that with `"another_target_name"` i.e. `"another_target_name"` is `"a"`. So `substitute` seems to be a bit redundant there?

**Reply by wlandau** [reply]
Yes, if `a` is the target name and you know it in advance, then you can just write `quote(f(a))`. `substitute(f(a), env = list())` gives the same result, but be careful if you do not set `env`.

**Reply by liutiming** [reply]
thanks! how about `pattern`?

**Reply by wlandau** [reply]
`pattern` and `command` are both expression/language objects, and the metaprogramming is the same.

**Reply by liutiming** [reply]
Thank you so much! I am getting the hang of it and I realise it is a very monotonous (and error prone) process... Really appreciate the work you have done for STAN and JAGS!
